packages/workflow-core/src/lib/utils/deep-merge-with-options.ts

```diff
@@ -0,0 +1,139 @@
+import { isObject } from '@ballerine/common';
+
+type UnknownRecord = Record<PropertyKey, unknown>;
+
+const mergeObjects = (obj1: UnknownRecord, obj2: UnknownRecord) => {
+  const result = { ...obj1 };
+
+  for (let key in obj2) {
+    if (isObject(obj2[key]) && !Array.isArray(obj2[key]) && key in result) {
+      result[key] = mergeObjects(result[key] as UnknownRecord, obj2[key] as UnknownRecord);
+    } else {
+      result[key] = obj2[key];
+    }
+  }
+
+  return result;
+};
```

Comment:
you'll modify the obj1 inner fields,

```suggestion
const mergeObjects = (obj1: UnknownRecord, obj2: UnknownRecord): UnknownRecord => {
  const result: UnknownRecord = {};

  for (let key in obj1) {
    if (isObject(obj1[key]) && !Array.isArray(obj1[key]) && key in obj2) {
      result[key] = mergeObjects(obj1[key] as UnknownRecord, obj2[key] as UnknownRecord);
    } else {
      result[key] = obj1[key];
    }
  }

  for (let key in obj2) {
    if (!(key in result)) {
      result[key] = obj2[key];
    }
  }

  return result;
};
```